### PR TITLE
[YoutubeDL] Make `bestvideo+bestaudio/best` default format when merger is available

### LINKED ISSUE #5447: Document and make default --format='bestvideo+bestaudio/best'
Please document ```--format='bestvideo+bestaudio/best'``` in man page and consider making it default.

### PR DESCRIPTION
**Pull request by dstftw**
Closes #5447.

Pros:
+ Results in best overall quality.

Cons:
- Potential merge issues (#5298).
- Unable to watch video while downloading (by piping to video player).

### CODE REVIEW

**Comment by jaimeMF** [comment]
In some issue report someone commented saying that for videos in low resolution it's more convenient to just get the normal formats, since they have the same or better quality. Checking these cases would probably complicate the code too much and I don't think it's worth the effort.

> Potential merge issues

We should make sure that the default formats work (otherwise users would get confused), I propose two alternatives:
* If bestaudio and bestvideo are not compatible, we merge into a mkv file
* Use `bestvideo[mp4]+bestaudio[m4a]/bestvideo[webm]+bestaudio[webm]/mp4` instead, with this option we may not get the best quality (I think in some cases there's either a webm or mp4 with better resolution, but I'm not really sure).

With this change the YoutubeIE test fails, we have to change the `format` field in `test/parameters.json`.

Apart from the previous comments, I like the change and agree that `bestvideo+bestaudio/best` should be the default

**Comment by dstftw** [comment]
I think merging in mkv sounds reasonable since theoretically there may be a situation when there are no video/audio of the compatible formats available. Moreover, general video+audio format selection will benefit from it as well.

**Comment by jaimeMF** [comment]
I also prefer defaulting to mkv. Some players may not support it, I don't think it's a big problem (it happens the same if some website only gives flv files), but some users may not like it.

We could also merge into mkv and use `bestvideo[mp4]+bestaudio[m4a]/bestvideo[webm]+bestaudio[webm]/bestvideo+bestaudio/mp4`, but that's quite ugly, I'd prefer to keep it as simple as possible.

File: youtube_dl/YoutubeDL.py [hunk 1]
```diff
@@ -1369,7 +1373,29 @@ def dl(name, info):
...
+                                if video_ext in exts and audio_ext in exts:
+                                    return True
+                        # TODO: Check acodec/vcodec
+                        return False
+
+                    requested_formats = info_dict['requested_formats']
+                    # Merge incompatible formats into mkv
+                    if not compatible_formats(requested_formats):
+                        filename = os.path.splitext(filename)[0] + '.mkv'
+                        self.report_warning('You have requested formats uncompatible for merge. '
```
Comment (jaimeMF) [inline youtube_dl/YoutubeDL.py]: Since `bestvideo+bestaudio/best` will be the default, maybe it's better to use `to_screen`.

**Comment by jaimeMF** [comment]
Before merging this I think we should detect if the video has already been downloaded and merged. Currently if you don't use --keep-video, the files are redownloaded and merged again (if you prefer, I can try to implement it myself).

One issue with merging is that for long videos (like fEAyDNWaCto from #5298), ffmpeg takes a relatively long time to merge them. I don't know if we can do something to fix it.

File: youtube_dl/YoutubeDL.py (hunk 1 above)
Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: Why? Uncompatible formats scenario is still expected to happen less often that compatible and can be treated as *reasonable-to-warn-about*.

**Comment by dstftw** [comment]
I've added basic check for already merged file. Feel free to tweak it if that's not what you expected.

File: youtube_dl/YoutubeDL.py (hunk 1 above)
Comment (jaimeMF) [inline youtube_dl/YoutubeDL.py]: You are right, it's fine.

**Comment by jaimeMF** [comment]
I have written an alternative in #5469, we have to use a temporary file with ffmpeg, otherwise we can end with a partial video if we cancel the merging.
(If you like it we can merge it to master directly or you cherry pick it in your branch, as you prefer)

**Comment by dstftw** [comment]
Ok, lets just merge both to master.

2015-04-19 21:03 GMT+06:00 Jaime Marquínez Ferrándiz <
notifications@github.com>:

**Comment by jaimeMF** [comment]
Cherry picked everything except the last commit (feccf29c876869f44a9c983977371073b9801a51, 81cd954a512e48fb5cccc3159cd0581088bff0b3 and 70947ea7b13d22a55756dc2b6d086e058edd91b1). Nice work!

**Pull request closed without merging**